net/captivedetection/captivedetection.go

```diff
@@ -0,0 +1,189 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+// Package captivedetection provides a way to detect if the system is connected to a network that has
+// a captive portal. It does this by making HTTP requests to known captive portal detection endpoints
+// and checking if the HTTP responses indicate that a captive portal might be present.
+package captivedetection
+
+import (
+	"context"
+	"log"
+	"net"
+	"net/http"
+	"net/url"
+	"strings"
+	"sync"
+	"syscall"
+	"time"
+
+	"tailscale.com/net/netmon"
+	"tailscale.com/tailcfg"
+	"tailscale.com/types/logger"
+)
+
+// Detective is a type used to detect if the system is behind a captive portal.
+type Detector struct {
+	// mu is the mutex used to ensure that only one captive portal detection attempt is running at a time.
+	mu sync.Mutex
+	// httpClient is the HTTP client that is used for captive portal detection. It is configured to not follow redirects
+	// and to have a short timeout since it will talk to a LAN device.
+	httpClient *http.Client
+}
+
+func NewDetector() *Detector {
+	return &Detector{
+		httpClient: &http.Client{
+			// No redirects allowed
+			CheckRedirect: func(req *http.Request, via []*http.Request) error {
+				return http.ErrUseLastResponse
+			},
+			Transport: http.DefaultClient.Transport,
+			Jar:       http.DefaultClient.Jar,
+			Timeout:   Timeout,
+		},
+	}
+}
+
+// Timeout is the timeout for captive portal detection requests. Because the captive portal intercepting our requests
+// is usually located on the LAN, this is a very short timeout.
+const Timeout time.Duration = 500 * time.Millisecond
+
+// DetectCaptivePortal is the entry point to the API. It attempts to detect if the system is behind a captive portal
+// by making HTTP requests to known captive portal detection Endpoints. If any of the requests return a response code
+// or body that looks like a captive portal, we return true. We return false in all other cases, including when any
+// error occurs during a detection attempt.
+//
+// This function might take a while to return, as it will attempt to detect a captive portal on all available interfaces
+// by performing multiple HTTP requests. It should be called in a separate goroutine if you want to avoid blocking.
+func (d *Detector) DetectCaptivePortal(ifst *netmon.State, derpMap *tailcfg.DERPMap, preferredDERPRegionID int, logf logger.Logf, goos string) (found bool, err error) {
```

Comment:
Bit of a nitpick, but you could split this into a wrapper function; callers from other packages will presumably not need to set the `goos` argument, so you could do:

```suggestion
func (d *Detector) DetectCaptivePortal(ifst *netmon.State, derpMap *tailcfg.DERPMap, preferredDERPRegionID int, logf logger.Logf) (found bool, err error) {
	return d.detectCaptivePortalWithGOOS(ifst, derpMap, preferredDERPRegionID, logf, runtime.GOOS)
}

func (d *Detector) detectCaptivePortalWithGOOS(ifst *netmon.State, derpMap *tailcfg.DERPMap, preferredDERPRegionID int, logf logger.Logf, goos string) (found bool, err error) {
```

Reply:
Done